rfd/0152-automatic-database-users-mongodb.md

```diff
@@ -0,0 +1,300 @@
+    { "resource": { "db": "$external", "collection": "" }, "actions": [ "createUser", "updateUser", "dropUser", "viewUser", "setAuthenticationRestriction", "changeCustomData"] }
+  ],
+  "roles": []
+}
+```
+Where:
+- `inprog` action is required to run `currentOp` for searching active
+  connections.
+- `grantRole` and `revokeRole` actions are required to manage roles on all
+  databases.
+- User related actions are limited on database `$external` database as X.509
+  users only exist on `$external`.
+
+The admin user must be created on `$external` database with X.509 authentication:
+```json
+{
+    "createUser": "CN=teleport-admin",
+    "roles": [ {"role": "teleport-admin-role", "db": "admin"} ]
+}
+```
+
+As the implementation of other databases, the name of admin user is
+defined in the database spec `admin_user.name`.
+
+However, there is NO concept of "Stored Procedures" in MongoDB. The user
+provisioning logic will be carried through multiple `runCommand` calls
+implemented in Go. Multiple parallel database sessions will not race thanks to
+[semaphore
+locking](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md#locking).
+
+### Roles
+
+Unlike MySQL or PostgreSQL where roles are scoped to the entire database
+instance/cluster, a role in MongoDB is scoped to a specific database.
+
+For example, a custom role `myCustomRole` can be created on database `db1` with
+specific privileges on `db1`, and another role `myCustomRole` can be created on
+database `db2` with specific privileges on `db2`. They will be considered two
+different roles.
+
+Most built-in roles are available on all databases, while some built-in roles
+like `readAnyDatabase` can only be applied on the `admin` database.
+
+Therefore, when specifying database roles to assign for the user, it must be in
+the format of `<role-name>@<db-name>` to fully identify the role:
+
+```yaml
+kind: "role"
+version: "v6"
+metadata:
+  name: "example"
+spec:
+  options:
+    create_db_user_mode: keep
+  allow:
+    db_names:
+    - "db1"
+    - "db2"
+    - "db3"
+    db_roles: 
+    - "readAnyDatabase@admin"
+    - "readWrite@db2"
+    - "myCustomRole@db3"
+```
+
+Teleport assigns the roles specified in `db_roles` to the auto-provisioned user
+and the MongoDB cluster restricts in-database access based on the assigned
+roles. On top of that, Teleport enforces that the user can only access
+databases listed in `db_names`.
+
+For example, in the above sample role, even though the user is assigned role
+`readAnyDatabase@admin`, Teleport will block access to databases not in the
+`db_names`.
+
+Of course, the user has the option to use `*` for `db_names` to soly rely on
+MongoDB's role management to restrict access.
+
+### User accounts
+
+New users with name `CN=<teleport-username>` will be created on `$external`
+database to use X.509 authentication.
+
+All auto-provisioned users will have the following `customData` to indicate the
```

Comment:
I know I suggested this, but we have to be careful not to override any custom data that might've been set by users. It's just a json object, right?

Reply:
yes. We set `"customData": {"teleport-auto-user": true}` at `createUser`. We don't provide `customData` in `updateUser` calls which preserve the existing value of `customData` in case the user adds something. Added a note in the RFD. 

Though ideally, no other database users should change this `customData` as it's managed by Teleport. It won't be preserved in `drop` mode either.